### PR TITLE
Logger

### PR DESCRIPTION
**Pull request by AviadSofer**
Hi
add kind of Logger to the Repo.
not finel yet:)

### CODE REVIEW

File: src/services/storage/index.ts [hunk 1]
```diff
@@ -37,19 +39,25 @@ export class Storage implements StorageDriver {
...
 			return this.driver.url.deleteOverdue(timespanMs)
 		}
 		public async edit(id: string, url: string): Promise<StoredUrl> {
+			logger.debug(`Running edit with ${id} and ${url}`)
 			return this.driver.url.edit(id, url)
 		}
 
 		public async save(url: string): Promise<StoredUrl> {
+			// logger.debug(`Start save with ${url}`)
+			// סליחה על העברית, התוצאה שמוחזרת היא לכאורה רגישה, אז כרגע זה בקומיט
```
Comment (SnirShechter) [inline src/services/storage/index.ts:60]: This is actually not sensitive. You can remove the Hebrew and add the logger.

**What's sensitive?**
Database credentials, admin token, personal information (which we currently don't have).

**What's _not_ sensitive?**
ids, URLs (without tokens in the query etc.), IPs and anything else.

**Review by SnirShechter — CHANGES_REQUESTED** [review]

File: src/config/validate.ts [hunk 2]
```diff
@@ -2,15 +2,21 @@ import ms from 'ms'
...
 	validateBaseRedirectUrl(rawConfig.baseRedirectUrl, rawConfig.apiPrefix)
 	validateStorageDriver(rawConfig.storage)
 	validateUrlLifetime(rawConfig.url.lifetime)
+	validateLevelLog(rawConfig.logLevel)
 
 	return true
 }
 function validateStorageDriver(storage: RawConfig['storage']): void {
+	const storageValues = Object.values(storage)
+	logger.debug(`Start validateStorageDriver with ${storageValues}`)
```
[the remark this answers is not in the document]
Comment (AviadSofer) [reply src/config/validate.ts:19]: ok:)
actually, I tried. if you print just `storage ` itself, it's will print just "Obj"

File: src/services/logger/logger.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+const logger = Pino({
+	prettyPrint: true,
+})
+
+logger.setLevel = function setLevel(newLevel: string) {
+	logger.level = newLevel
+}
+
+logger.on('level-change', () => {
+	console.log(`Logger: Level changed to ${logger.level}`)
```
[the remark this answers is not in the document]
Comment (AviadSofer) [reply src/services/logger/logger.ts:11]: Hi.
If we will change this to debug, while the level define as a "warn" for exemple, it's will not print anything.
You can see this [here](https://github.com/pinojs/pino/blob/master/docs/api.md#event-level-change).

File: src/config/index.ts [hunk 4]
```diff
@@ -4,10 +4,12 @@ import { AuthDriverName } from '../services/auth/types/config.js'
 import { normalizeConfig } from './normalize.js'
 import type { RawConfig, Config } from './types.js'
 import { validateConfig } from './validate.js'
+import { logger } from '../services/logger/logger.js'
```
Comment (SnirShechter) [inline src/config/index.ts:7]: Remove this (unused import)

File: src/config/index.ts [hunk 5]
```diff
@@ -36,15 +38,18 @@ const rawConfig: RawConfig = {
...
+	    A default BearerToken driver is selected and a random bearer token will be generated.`)
 
 	rawConfig.auth.driverName = AuthDriverName.BearerToken
 	rawConfig.auth.bearerTokenDriverConfig.token = cryptoRandomString({ length: 18, type: 'alphanumeric' })
 	console.info(`Generated bearer token is '${rawConfig.auth.bearerTokenDriverConfig.token}'`)
 }
 
-if (!validateConfig(rawConfig)) throw new InvalidConfigError()
+if (!validateConfig(rawConfig)) {
+	logger.error('rawConfig is not validated')
```
Comment (SnirShechter) [inline src/config/index.ts:50]:
```suggestion
	logger.error('Recieved config is invalid')
```

These logs will be seen by users. They don't know what RawConfig means.

File: src/index.ts [hunk 6]
```diff
@@ -13,10 +14,14 @@ declare module 'fastify' {
 	}
 }
 
+// Logger
+logger.info(`Logger level define as ${config.logLevel}`)
```
Comment (SnirShechter) [inline src/index.ts:18]:
```suggestion
logger.info(`Logger level defined as ${config.logLevel}`)
```

**Review by SnirShechter — CHANGES_REQUESTED** [review]

File: src/config/validate.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (SnirShechter) [reply src/config/validate.ts]: You can print it as `logger.debug(`...`, storage)

File: src/services/storage/index.ts [hunk 7]
```diff
@@ -37,20 +45,50 @@ export class Storage implements StorageDriver {
...
 		}
 
 		public async save(url: string): Promise<StoredUrl> {
-			return this.driver.url.save(url)
+			try {
+				logger.debug(`Start Storage.url.save with ${url}`)
+				return this.driver.url.save(url)
+			} catch (err) {
+				logger.error(`Storage.url.save failed: ${err}`)
+				throw new GeneralError('Could not save (Storage.url.save)')
```
Comment (SnirShechter) [inline src/services/storage/index.ts:90]: Remember that these errors are returned to the client, so don't add the function name here.
```suggestion
				throw new GeneralError('Could not save url')
```

File: src/services/storage/index.ts [hunk 8]
```diff
@@ -28,7 +30,13 @@ export class Storage implements StorageDriver {
 	}
 	public async initialize(): Promise<void> {
 		// Waits for 1 minute (6 * 10,000ms) before failing
-		await runWithRetries(this._driver.initialize.bind(this._driver), { retries: 6, retryTime: 10 * 1000 })
+		try {
+			logger.debug(`Running Storage.initialize`)
+			await runWithRetries(this._driver.initialize.bind(this._driver), { retries: 6, retryTime: 10 * 1000 })
+		} catch (err) {
+			logger.error(`Storage.initialize failed: ${err}`)
+			throw new GeneralError('Could not initialize (Storage.initialize)')
```
Comment (SnirShechter) [inline src/services/storage/index.ts:38]: Can you throw the actual error here? This function is not propagated back to the client, so it's safe to throw it
```suggestion
			throw err
```

File: src/services/storage/index.ts [hunk 9]
```diff
@@ -37,20 +45,50 @@ export class Storage implements StorageDriver {
 			return this.storage._driver
 		}
 		public async get(id: string): Promise<StoredUrl> {
-			return this.driver.url.get(id)
+			try {
+				logger.debug(`Running Storage.url.get with ${id}`)
+				return this.driver.url.get(id)
+			} catch (err) {
+				logger.error(`Storage.url.get failed: ${err}`)
```
Comment (SnirShechter) [inline src/services/storage/index.ts:52]: Did you check that these logs actually show the error well? If they do, just ignore this comment.

File: src/services/auth/index.ts [hunk 10]
```diff
@@ -25,9 +26,11 @@ export class Auth implements AuthDriver {
 	}
 
 	public async isAuthorized(request: FastifyRequest): Promise<boolean> {
+		logger.debug(`Running isAuthorized`)
```
Comment (SnirShechter) [inline src/services/auth/index.ts:29]:
```suggestion
		logger.debug(`Running Auth.isAuthorized`)
```

File: src/services/auth/index.ts [hunk 11]
```diff
@@ -25,9 +26,11 @@ export class Auth implements AuthDriver {
...
 		return this._driver.isAuthorized(request)
 	}
 	public authorize(request: FastifyRequest): Promise<void> {
+		logger.debug(`Running authorize`)
```
Comment (SnirShechter) [inline src/services/auth/index.ts:33]:
```suggestion
		logger.debug(`Running Auth.authorize`)
```

File: src/services/auth/drivers/bearerToken/index.ts [hunk 12]
```diff
@@ -18,11 +20,13 @@ export class BearerTokenAuth implements AuthDriver {
 		return token
 	}
 	public async isAuthorized(request: FastifyRequest): Promise<boolean> {
+		logger.debug(`Running isAuthorized`)
 		const token = this.verifyHeaderAndExtractToken(request)
 
 		return token !== null && token === this.token
 	}
 	public async authorize(request: FastifyRequest): Promise<void> {
+		logger.debug(`Running authorize`)
```
Comment (SnirShechter) [inline src/services/auth/drivers/bearerToken/index.ts:29]:
```suggestion
		logger.debug(`Running BearerTokenAuth.authorize`)
```

File: src/services/auth/drivers/bearerToken/index.ts [hunk 13]
```diff
@@ -18,11 +20,13 @@ export class BearerTokenAuth implements AuthDriver {
 		return token
 	}
 	public async isAuthorized(request: FastifyRequest): Promise<boolean> {
+		logger.debug(`Running isAuthorized`)
```
Comment (SnirShechter) [inline src/services/auth/drivers/bearerToken/index.ts:23]:
```suggestion
		logger.debug(`Running BearerTokenAuth.isAuthorized`)
```

File: src/index.ts (hunk 6 above)
Comment (SnirShechter) [inline src/index.ts:18]:
```suggestion
logger.info(`Logger level defined as ${config.logLevel}`)
```

File: src/config/index.ts [hunk 14]
```diff
@@ -36,15 +38,18 @@ const rawConfig: RawConfig = {
...
 
 if (!rawConfig.auth.driverName) {
-	console.warn(`No auth driver selected.
-		A default BearerToken driver is selected and a random bearer token will be generated.`)
+	logger.warn(`No auth driver selected.
+	    A default BearerToken driver is selected and a random bearer token will be generated.`)
 
 	rawConfig.auth.driverName = AuthDriverName.BearerToken
 	rawConfig.auth.bearerTokenDriverConfig.token = cryptoRandomString({ length: 18, type: 'alphanumeric' })
 	console.info(`Generated bearer token is '${rawConfig.auth.bearerTokenDriverConfig.token}'`)
```
Comment (SnirShechter) [inline src/config/index.ts:46]:
```suggestion
	logger.info(`Generated bearer token is '${rawConfig.auth.bearerTokenDriverConfig.token}'`)
```

**Review by SnirShechter — CHANGES_REQUESTED** [review]

**Comment by SnirShechter** [comment]
Closes #21
Closes #80

**Review by SnirShechter — APPROVED** [review]

**Pull request merged**